rust/ccommon_rs/src/option.rs

```diff
@@ -0,0 +1,577 @@
+
+    /// Print out the values of all options.
+    ///
+    /// Internally this calls out to `option_print_all`.
+    fn print_all(&self) {
+        unsafe {
+            option_print_all(
+                // Note: ccommon uses a mutable pointer but it
+                //       should really be a const pointer.
+                self.as_ptr() as *mut _,
+                Self::num_options() as u32,
+            )
+        }
+    }
+
+    /// Load default values for all options.
+    ///
+    /// Internally this calls `option_load_default`
+    fn load_default(&mut self) -> Result<(), crate::Error> {
+        let status = unsafe { option_load_default(self.as_mut_ptr(), Self::num_options() as u32) };
+
+        if status == 0 {
+            Ok(())
+        } else {
+            Err(status.into())
+        }
+    }
+
+    /// Load options from a file in `.ini` format.
+    ///
+    /// Internally this calls `option_load_file`.
+    fn load_from_libc_file(&mut self, file: *mut libc::FILE) -> Result<(), crate::Error> {
+        let status =
+            unsafe { option_load_file(file, self.as_mut_ptr(), Self::num_options() as u32) };
+
+        if status == 0 {
+            Ok(())
+        } else {
+            Err(status.into())
+        }
+    }
+}
+
+impl<T: Options> OptionExt for T {}
+
+/// A boolean option.
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct Bool(option);
+
+impl Sealed for Bool {}
+
+unsafe impl SingleOption for Bool {
+    type Value = bool;
+
+    fn new(default: bool, name: &'static CStr, desc: &'static CStr) -> Self {
+        Self(option {
+            name: name.as_ptr() as *mut _,
+            set: false,
+            type_: OPTION_TYPE_BOOL,
+            default_val: option_val_u { vbool: default },
+            val: option_val_u { vbool: default },
+            description: desc.as_ptr() as *mut _,
+        })
+    }
+    fn defaulted(name: &'static CStr, desc: &'static CStr) -> Self {
+        Self::new(Default::default(), name, desc)
+    }
+
+    fn name(&self) -> &'static CStr {
+        unsafe { CStr::from_ptr(self.0.name) }
+    }
+    fn desc(&self) -> &'static CStr {
+        unsafe { CStr::from_ptr(self.0.description) }
+    }
+    fn value(&self) -> Self::Value {
+        unsafe { self.0.val.vbool }
+    }
+    fn default(&self) -> Self::Value {
+        unsafe { self.0.default_val.vbool }
+    }
+    fn is_set(&self) -> bool {
+        self.0.set
+    }
+
+    fn set_value(&mut self, val: Self::Value) {
+        self.0.set = true;
+        self.0.val = option_val_u { vbool: val }
+    }
+}
+
+impl fmt::Debug for Bool {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Bool")
+            .field("name", &self.name())
+            .field("desc", &self.desc())
+            .field("value", &self.value())
+            .field("default", &self.default())
+            .field("is_set", &self.is_set())
+            .finish()
+    }
+}
+
+/// An unsigned integer option.
```

Comment:
I wonder if splitting the individual structs out into submodules would be better.